### ISSUE TITLE
TypeVar's infer_variance setting seems broken

### ISSUE THREAD

**Issue by rezzubs** [microsoft/pyright#11068]
## Overview

I was looking into how variance works with 3.12's generic type parameters and it looks like it should be inferred. This can be replicated by setting `infer_variance=True` for a manually created `TypeVar`. Neither seems to have the desired effect. I will demonstrate this by setting the `covariant` and `contravariant` parameters of a `TypeVar` and comparing the outputs to the inferred version. Pyright was ran from zsh on macOS 26.

Environment:
```
python --version
Python 3.13.5

pyright --version
pyright 1.1.407
```

## Covariant cases

The outputs are expected to match for both

<details>
<summary>Code `covariant=True`</summary>

```python
from dataclasses import dataclass
from typing import TypeVar, Generic


class Base: ...


class Super(Base): ...


T = TypeVar("T", covariant=True)


@dataclass
class Box(Generic[T]):
    val: T

    def get_val(self) -> T:
        return self.val

    # def set_val(self, val: T):
    #     self.val = val


base: Box[Base] = Box(Base())
super: Box[Super] = Box(Super())


def use_base(base: Box[Base]):
    _ = base


def use_super(super: Box[Super]):
    _ = super


use_base(base)
use_base(super)

use_super(base)
use_super(super)
```
</details>

<details>
<summary>Output for manually setting `covariant=True`</summary>

```
/Users/rezzubs/dev/variance-test/main.py
  /Users/rezzubs/dev/variance-test/main.py:40:11 - error: Argument of type "Box[Base]" cannot be assigned to parameter "super" of type "Box[Super]" in function "use_super"
    "Box[Base]" is not assignable to "Box[Super]"
      Type parameter "T@Box" is covariant, but "Base" is not a subtype of "Super"
        "Base" is not assignable to "Super" (reportArgumentType)
1 error, 0 warnings, 0 notes
```
</details>

<details>
<summary>Code for `infer_variance=True`</summary>

```python
from dataclasses import dataclass
from typing import TypeVar, Generic


class Base: ...


class Super(Base): ...


T = TypeVar("T", infer_variance=True)


@dataclass
class Box(Generic[T]):
    val: T

    def get_val(self) -> T:
        return self.val

    # def set_val(self, val: T):
    #     self.val = val


base: Box[Base] = Box(Base())
super: Box[Super] = Box(Super())


def use_base(base: Box[Base]):
    _ = base


def use_super(super: Box[Super]):
    _ = super


use_base(base)
use_base(super)

use_super(base)
use_super(super)
```
</details>

<details>
<summary>Output for `infer_variance=True`</summary>

```
/Users/rezzubs/dev/variance-test/main.py
  /Users/rezzubs/dev/variance-test/main.py:38:10 - error: Argument of type "Box[Super]" cannot be assigned to parameter "base" of type "Box[Base]" in function "use_base"
    "Box[Super]" is not assignable to "Box[Base]"
      Type parameter "T@Box" is invariant, but "Super" is not the same as "Base" (reportArgumentType)
  /Users/rezzubs/dev/variance-test/main.py:40:11 - error: Argument of type "Box[Base]" cannot be assigned to parameter "super" of type "Box[Super]" in function "use_super"
    "Box[Base]" is not assignable to "Box[Super]"
      Type parameter "T@Box" is invariant, but "Base" is not the same as "Super" (reportArgumentType)
2 errors, 0 warnings, 0 notes
```
</details>

Notice how it labels `T@Box` as invariant.

## Contravariant cases

The outputs are expected to match for both

<details>
<summary>Code for `contravariant=True</summary>

```python
from dataclasses import dataclass
from typing import TypeVar, Generic


class Base: ...


class Super(Base): ...


T = TypeVar("T", contravariant=True)


@dataclass
class Box(Generic[T]):
    val: T

    # def get_val(self) -> T:
    #     return self.val

    def set_val(self, val: T):
        self.val = val


base: Box[Base] = Box(Base())
super: Box[Super] = Box(Super())


def use_base(base: Box[Base]):
    _ = base


def use_super(super: Box[Super]):
    _ = super


use_base(base)
use_base(super)

use_super(base)
use_super(super)
```
</details>

<details>
<summary>Output for `contravariant=True`</summary>

```
/Users/rezzubs/dev/variance-test/main.py
  /Users/rezzubs/dev/variance-test/main.py:38:10 - error: Argument of type "Box[Super]" cannot be assigned to parameter "base" of type "Box[Base]" in function "use_base"
    "Box[Super]" is not assignable to "Box[Base]"
      Type parameter "T@Box" is contravariant, but "Super" is not a supertype of "Base"
        "Base" is not assignable to "Super" (reportArgumentType)
1 error, 0 warnings, 0 notes
```
</details>

<details>
<summary>Code for `infer_variance=True`</summary>

```python
from dataclasses import dataclass
from typing import TypeVar, Generic


class Base: ...


class Super(Base): ...


T = TypeVar("T", infer_variance=True)


@dataclass
class Box(Generic[T]):
    val: T

    # def get_val(self) -> T:
    #     return self.val

    def set_val(self, val: T):
        self.val = val


base: Box[Base] = Box(Base())
super: Box[Super] = Box(Super())


def use_base(base: Box[Base]):
    _ = base


def use_super(super: Box[Super]):
    _ = super


use_base(base)
use_base(super)

use_super(base)
use_super(super)
```
</details>

<details>
<summary>Output for `infer_variance=True`</summary>

```
/Users/rezzubs/dev/variance-test/main.py
  /Users/rezzubs/dev/variance-test/main.py:38:10 - error: Argument of type "Box[Super]" cannot be assigned to parameter "base" of type "Box[Base]" in function "use_base"
    "Box[Super]" is not assignable to "Box[Base]"
      Type parameter "T@Box" is invariant, but "Super" is not the same as "Base" (reportArgumentType)
  /Users/rezzubs/dev/variance-test/main.py:40:11 - error: Argument of type "Box[Base]" cannot be assigned to parameter "super" of type "Box[Super]" in function "use_super"
    "Box[Base]" is not assignable to "Box[Super]"
      Type parameter "T@Box" is invariant, but "Base" is not the same as "Super" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```
</details>

Again, `T@Box` is labeled as invariant.

**Comment by rezzubs** [comment]
Looks like I messed up the naming 🙃. What I called `Super` should probably be called `Sub` as it's the subclass of `Base` but I digress

**Comment by erictraut** [comment]
I think pyright is working correctly here. You're using `@dataclass` classes, which (as of Python 3.13) synthesize a `__replace__` method whose signature results in all class-scoped type variables to be inferred as invariant. There has been [significant discussion](https://discuss.python.org/t/make-replace-stop-interfering-with-variance-inference/96092) about this already, but the typing community has not yet achieved consensus on how to address it.

If you use a non-dataclass, you'll see that `infer_variance` works as you expect.
